File: my-folder/0103-binary-tree-zigzag-level-order-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        result = []
        queue = deque([root])
        level = 0

        while queue:
            level_size = len(queue)
            level_nodes = []

            for _ in range(level_size):
                node = queue.popleft()

                if level % 2 == 0:
                    level_nodes.append(node.val)
                else:
                    level_nodes.insert(0, node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(level_nodes)
            level += 1

        return result

File: my-folder/0103-binary-tree-zigzag-level-order-traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        boolean isLeft = true;
        Queue<TreeNode> d = new LinkedList<>();
        d.offer(root);
        while (!d.isEmpty()) {
            int size = d.size();
            List<Integer> temp = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = d.poll();
                if (isLeft) {
                    temp.add(node.val);
                }
                else {
                    temp.add(0, node.val);
                }
                if (node.left != null) {
                    d.add(node.left);
                }
                if (node.right != null) {
                    d.add(node.right);
                }
            }
            res.add(temp);
            isLeft = !isLeft;
        }
        return res;
    }
}
